### ISSUE TITLE
Exporting tflite version for use in Android - Missing operations

### ISSUE THREAD

**Issue by kvaggelakos** [ultralytics/yolov5#5599]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Export

## Preface
This may not be a suitable bug report for this repo, but I'm filing it here anyway in case others are seeing the same issue. 
I know @zldrobit has been working on an Android implementation using this repo, maybe he can chime in.

## Issue

I tried exporting my custom yolov5  model using the export.py, which I called like this:
```
python export.py --weights ./runs/train/yolo_proj/weights/best.pt --include tflite
```

This is the error I'm seeing using **tensorflow 2.4.0**:
```
java.lang.IllegalArgumentException: Internal error: Failed to apply delegate: Following operations are not supported by GPU delegate:
DEQUANTIZE: 
389 operations will run on the GPU, and the remaining 1 operations will run on the CPU.
Can not open OpenCL library on this device - dlopen failed: library "libOpenCL.so" not found
Falling back to OpenGL
TfLiteGpuDelegate Init: No shader implementation for transpose
TfLiteGpuDelegate Prepare: delegate is not initialized
Node number 390 (TfLiteGpuDelegateV2) failed to prepare.

Restored original ex
```

This is the error I'm seeing using **tensorflow 2.5.0**:
```
java.lang.IllegalArgumentException: Internal error: Failed to apply delegate: Can not open OpenCL library on this device - dlopen failed: library "libOpenCL.so" not found
Falling back to OpenGL
TfLiteGpuDelegate Init: No shader implementation for transpose
TfLiteGpuDelegate Prepare: delegate is not initialized
Node number 390 (TfLiteGpuDelegateV2) failed to prepare.

Restored original execution plan after delegate application failure.
```

### Environment

- Yolov5 latest master
- Using a pixel 6

### Additional

Not sure if this is related but I'm getting this error when I'm exporting the model:
```
2021-11-09 20:18:46.600920: E tensorflow/stream_executor/cuda/cuda_driver.cc:271] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by zldrobit** [comment]
@kvaggelakos Considering TFLite is moving from OpenGL to OpenCL and Pixel 6 is a very new type of cell phone, it's a surprise that OpenCL does not exist in the environment. I also found an issue discussing Google Pixel Phones does not support OpenCL library https://github.com/mapbox/mapbox-vision-android/issues/256. Another issue in TensorFlow https://github.com/tensorflow/tensorflow/issues/44124#issuecomment-712437075 discusses Transpose operation, yet with no obvious answer.  

I could reproduce this problem with the TFLite benchmark tool:
![image](https://user-images.githubusercontent.com/3896992/141244169-fc00e2a9-4482-4f84-8fb3-ede315ce090c.png)

The root cause is the Transpose operation is not supported by OpenGL GPU delegate.

**Comment by kvaggelakos** [comment]
@zldrobit Thanks for the answer.

Do you think the pixel 6 would require to run TF with OpenGL or is there another way? Something I fail to understand is how the tensor chip can be used with TF. To me it seems weird that OpenGL is the way to access that compute, but I don't have enough knowledge. I'll keep the issue open for now and report back if I find some answers.

**Comment by zldrobit** [comment]
@kvaggelakos Yes, you could run TF with NNAPI. From https://www.anandtech.com/show/17032/tensor-soc-performance-efficiency/5, the most efficient way to run TF is with NNAPI. Neural network ops, e.g. convolutions, can be translated to the rendering language of OpenGL, so deep learning models could be inferred on TFLite using OpenGL. OpenGL only works with GPUs while some CPUs support OpenCL, e.g. Intel. TFLite GPU delegate first tries to load OpenCL for inference. If this fails, it loads OpenGL instead.

PS: By now, I don't have an Android phone supporting NNAPI.

**Comment by kvaggelakos** [comment]
@zldrobit Sorry for the late reply. I've been working on getting the input data in place. I haven't forgotten about your reply. Once I get to the point where I can test out the NNAPI I will report back.

**Comment by kvaggelakos** [comment]
Ok, I got it running using the NNAPI. It's unclear however which hw device that's actually being used. My observed latency on the pixel 6 was around 800ms, which is just not fast enough. I've moved to using `mobiledet` (https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/tf1_detection_zoo.md#pixel4-edge-tpu-models) from TF instead, a step I didn't want to take, but it has way better latency (3-4ms), being able to run on the TPU in the pixel 6. Accuracy is worse, but that's the trade off.

**Comment by zldrobit** [comment]
@kvaggelakos Thanks for sharing the information. The latency of mobiledet is very impressive, and I guess the reason is searching the network architecture for Pixel Edge TPUs (https://ai.googleblog.com/2019/11/introducing-next-generation-on-device.html) so that the network could work entirely on a Pixel Edge TPU.

**Comment by zldrobit** [comment]
In case you're still interested in trying YOLOv5 on Android, there's a way to tackle the problem. According to [another thread](https://github.com/ultralytics/yolov5/pull/1127#issuecomment-1056038455), adding `<uses-native-library>` in the manifest should solve the problem. I just noticed that Pixel 6 is running Android 12, and the problem is related to the default access policy of Android 12. Without the `<uses-native-library>`, an App cannot be linked to OpenCL libs.

**Comment by shelper** [comment]
@zldrobit hey there, I had the same problem here when i try to run native binary (compiled from c++), as you mentioned adding `<uses-native-library>` in the manifest should solve the problem, but that is only for java. 

i checked the documentation for  `<uses-native-library>`  and it says when run native binary, this is enabled by default. 

and the benchmark tool `android_aarch64_benchmark_model` is a native binary tool, so if `<uses-native-library>`  is enabled by default, we shouldn't see the error running this tool. 

Just want to check if you have any suggestions or insights, thanks!

**Comment by zldrobit** [comment]
@shelper Ah, I don't have a mobile phone running Android 12 at hand. So I don't know more about this issue beyond the scope of the documentation. Maybe you could consult an experienced Android developer or ask about this question in [the TensorFlow issues](https://github.com/tensorflow/tensorflow).
